File: src/lib/providers/bioclaw-proxy.ts
// BioClaw-proxy provider.
//
// The desktop app does NOT ship an OpenRouter key. Users authenticate with
// their BioClaw email (existing /api/auth/send-otp + /api/auth/verify-otp
// flow) and we route chat completions through our SaaS, which appends the
// server-side OpenRouter key and pipes the SSE response back unchanged.
//
// Wire shape (defined by SaaS /api/desktop/chat/completions):
//   * Endpoint:  https://chat.bioclaw.tech/api/desktop/chat/completions
//   * Auth:      Cookie: bioclaw_session=<token>
//   * Request:   OpenAI-compatible {messages, model, tools, ...}
//   * Response:  OpenAI-style SSE — same chunks OpenRouter emits
//
// The path component intentionally ends in `/chat/completions` so the
// existing openai-compatible provider — which appends that suffix —
// produces the right final URL when we hand it `<base>/api/desktop` as
// the endpoint. Same body shape; we only change auth (cookie instead of
// bearer) and the base URL.

import type { Provider, ProviderEvent, ProviderStreamRequest } from './index';
import { getProvider, registerProvider } from './index';
import type { ProviderId } from '../agent/types';

const DEFAULT_SAAS_BASE_URL = 'https://chat.bioclaw.tech';

class BioClawProxyProvider implements Provider {
  readonly id: ProviderId = 'bioclaw-proxy';

  async *streamMessages(req: ProviderStreamRequest): AsyncIterable<ProviderEvent> {
    const explicit = req.model.endpoint;
    const base = (explicit && explicit.length > 0 ? explicit : DEFAULT_SAAS_BASE_URL).replace(
      /\/+$/,
      '',
    );
    const patched: ProviderStreamRequest = {
      ...req,
      model: {
        ...req.model,
        // The openai-compatible provider appends `/chat/completions`, so
        // we hand it `<base>/api/desktop` and the call lands on the SaaS
        // route `<base>/api/desktop/chat/completions`.
        endpoint: `${base}/api/desktop`,
        auth: {
          kind: 'cookie',
          ...(req.model.auth?.apiKey ? { apiKey: req.model.auth.apiKey } : {}),
          cookieName: req.model.auth?.cookieName ?? 'bioclaw_session',
          ...(req.model.auth?.headers ? { headers: req.model.auth.headers } : {}),
        },
      },
    };
    const compat = getProvider('openai-compatible');
    yield* compat.streamMessages(patched);
  }
}

registerProvider(new BioClawProxyProvider());

File: sidecar-rs/src/chat/provider/bioclaw_proxy.rs
//! BioClaw-proxy provider — what the desktop UI uses out of the box.
//!
//! The desktop app does NOT ship an OpenRouter / Anthropic key. Users
//! sign in with their BioClaw account, and chat completions are routed
//! through `chat.bioclaw.tech/api/desktop/chat/completions`, which
//! appends the server-side OpenRouter key and pipes the SSE response
//! back unchanged. Wire format is OpenAI-compatible by design — the
//! actual stream parsing is done by `openai_compat`; this module just
//! rewrites the endpoint + swaps bearer auth for the `bioclaw_session`
//! cookie.
//!
//! Mirrors `src/lib/providers/bioclaw-proxy.ts` (~50 lines TS,
//! similarly thin here).

use anyhow::Result;

use super::openai_compat::OpenAiCompat;
use super::{Provider, ProviderStream, ProviderStreamRequest};
use crate::chat::types::Auth;

const DEFAULT_SAAS_BASE_URL: &str = "https://chat.bioclaw.tech";

pub struct BioClawProxy;

impl Provider for BioClawProxy {
    fn id(&self) -> &'static str {
        "bioclaw-proxy"
    }

    async fn stream_messages(&self, mut req: ProviderStreamRequest) -> Result<ProviderStream> {
        // The openai-compatible client appends `/chat/completions` to
        // the endpoint we give it, so we hand it `<base>/api/desktop`
        // and the final URL is `<base>/api/desktop/chat/completions`.
        let base = req
            .model
            .endpoint
            .as_deref()
            .filter(|s| !s.is_empty())
            .unwrap_or(DEFAULT_SAAS_BASE_URL)
            .trim_end_matches('/')
            .to_string();
        req.model.endpoint = Some(format!("{base}/api/desktop"));

        // Promote whatever auth we have to a cookie if it isn't already
        // — the SaaS endpoint authenticates session-style.
        req.model.auth = match req.model.auth {
            Auth::Cookie { cookie_name, token } => Auth::Cookie { cookie_name, token },
            Auth::Bearer { api_key } => Auth::Cookie {
                cookie_name: "bioclaw_session".into(),
                token: api_key,
            },
            Auth::None => Auth::None,
        };

        OpenAiCompat.stream_messages(req).await
    }
}
